### PR TITLE
Bump Doobie from 0.2.x to 0.5.x

### PR DESCRIPTION
**Pull request by adelbertc**
- Move Doobie Param.many usage to use Fragments
- Since Doobie calls now return Cats data structures this PR also
  includes some data structure migrations, mostly NonEmptyList
  and Validat{ion, ed}
- Some shims between Scalaz and Cats are still in place to keep
  this PR from becoming too big, to be removed in a future PR

### CODE REVIEW

File: core/src/main/scala/storage/h2.scala [hunk 1]
```diff
@@ -223,17 +228,17 @@ final case class H2Storage(xa: Transactor[IO]) extends (StoreOp ~> IO) {
                            FROM PUBLIC.deployment_statuses AS x
                            WHERE x.deployment_id = d.id
                            ORDER BY x.id DESC
-                       )
-                     WHERE u.name = ${name}
-                     AND d.namespace_id = ${nsid}
-                     AND ds.state IN (${statusStrings: statusStrings.type})
-                  """.query[DeploymentRow].process.list
+                       )""" ++
+                  whereAnd(fr"u.name = ${name}", fr"d.namespace_id = ${nsid}", in(fr"ds.state", statusStrings))).
```
Comment (adelbertc) [inline core/src/main/scala/storage/h2.scala]: cc @tpolecat

File: core/src/main/scala/storage/h2.scala [hunk 2]
```diff
@@ -400,13 +412,10 @@ final case class H2Storage(xa: Transactor[IO]) extends (StoreOp ~> IO) {
           FROM PUBLIC.deployment_statuses AS x
           WHERE x.deployment_id = d.id
           ORDER BY x.id DESC
-        )
-      WHERE namespace_id = ${ns}
-          AND ds.state IN (${statusStrings: statusStrings.type})
-      ORDER BY id DESC
-    """
+        )""" ++ whereAnd(fr"namespace_id = ${ns}", in(fr"ds.state", statusStrings)) ++
```
Comment (adelbertc) [inline core/src/main/scala/storage/h2.scala]: cc @tpolecat

File: core/src/main/scala/storage/h2.scala [hunk 3]
```diff
@@ -428,16 +437,13 @@ final case class H2Storage(xa: Transactor[IO]) extends (StoreOp ~> IO) {
           ORDER BY x.id DESC
         )
       JOIN PUBLIC.units AS u ON u.id = d.unit_id
-      WHERE u.name = ${unitName}
-          AND namespace_id = ${ns}
-          AND ds.state IN (${statusStrings: statusStrings.type})
-      ORDER BY id DESC
-    """
+      """ ++ whereAnd(fr"u.name = ${unitName}", fr"namespace_id = ${ns}", in(fr"ds.state", statusStrings)) ++
+      fr"ORDER BY id DESC"
```
Comment (adelbertc) [inline core/src/main/scala/storage/h2.scala]: cc @tpolecat

File: core/src/main/scala/storage/h2.scala [hunk 4]
```diff
@@ -474,19 +480,17 @@ final case class H2Storage(xa: Transactor[IO]) extends (StoreOp ~> IO) {
...
         JOIN PUBLIC.deployments d
           ON u.id = d.unit_id
         JOIN PUBLIC.deployment_statuses s
           ON s.deployment_id = d.id
-        WHERE d.namespace_id = $nsid
-          AND s.state IN (${statusStrings: statusStrings.type})
-        ORDER BY u.name ASC
-     """.query[(UnitName, Version, String)].map { case (name, version, guid) =>
+    """ ++ whereAnd(fr"d.namespace_id = ${nsid}", in(fr"s.state", statusStrings)) ++
+    fr"ORDER BY u.name ASC")
```
Comment (adelbertc) [inline core/src/main/scala/storage/h2.scala]: cc @tpolecat

File: core/src/main/scala/storage/h2.scala [hunk 5]
```diff
@@ -762,22 +778,20 @@ final case class H2Storage(xa: Transactor[IO]) extends (StoreOp ~> IO) {
...
     val repoIds = repos.map(_.id)
-    implicit val reposParam: Param[repoIds.type] = Param.many(repoIds)
 
     // soft delete
     val query = sql"""
       UPDATE PUBLIC.repositories
       SET is_deleted = true
-      WHERE id IN (${repoIds: repoIds.type})
-    """
+    """ ++ whereAnd(in(fr"id", repoIds))
```
Comment (adelbertc) [inline core/src/main/scala/storage/h2.scala]: cc @tpolecat

File: core/src/main/scala/storage/h2.scala [hunk 6]
```diff
@@ -1042,25 +1054,26 @@ final case class H2Storage(xa: Transactor[IO]) extends (StoreOp ~> IO) {
...
+    """ ++ whereAnd(
+      fr"fu.name = ${sn.serviceType}",
+      fr"fu.version like ${likeVersion}",
+      fr"tb.namespace_id = ${ns}",
+      in(fr"fd_ds.state", statuses),
+      fr"fd_ds2.deployment_id IS NULL",
+      in(fr"td_ds.state", statuses),
+      fr"td_ds2.deployment_id IS NULL"
+    ) ++
+    fr"""
```
Comment (adelbertc) [inline core/src/main/scala/storage/h2.scala]: cc @tpolecat

File: core/src/main/scala/storage/h2.scala [hunk 7]
```diff
@@ -1074,12 +1087,13 @@ final case class H2Storage(xa: Transactor[IO]) extends (StoreOp ~> IO) {
...
-        AND u.version LIKE $likeVersion
-        AND ds2.deployment_id IS NULL
-    """.query[DeploymentRow].list.flatMap(_.traverse(deploymentFromRow _))
+    """ ++ whereAnd(
+      in(fr"ds.state", statuses),
+      fr"d.namespace_id = ${ns}",
+      fr"u.name = ${name}",
+      fr"u.version LIKE ${likeVersion}",
+      fr"ds2.deployment_id IS NULL"
+    )).query[DeploymentRow].to[List].flatMap(_.traverse(deploymentFromRow _))
```
Comment (adelbertc) [inline core/src/main/scala/storage/h2.scala]: cc @tpolecat

**Comment by adelbertc** [comment]
NOTE: Please review, but do not merge yet. Want to test a basic deployment workflow before merging, will update here once I do.

**Comment by adelbertc** [comment]
Tested on a K8s cluster, released a job, verified it deployed and ran, released a new version, verified old job was GC'd and expiration times worked as expected - ready for review!

**Comment by timperrett** [comment]
merge when ready @adelbertc

**Pull request merged**